Add new resource `argocd_application_set`

onematchfox opened the pull request:
This is a giant resource, so:
- I could well have missed some settings/attributes, and the tests definitely do not cover all use cases/configuration options
- I completely understand that a thorough review will take time - if it's even possible to thoroughly review such a large change at all

I have, however, tried to ensure that the tests do at least cover the default/standard use cases as provided in the ArgoCD documentation. So, hopefully, a more cursory review of the code should suffice. I would also suggest we highlight in the release notes that there is a significant possibility of issues with this resource, so users should not expect it to be perfect right from the start.

Closes #223 

Notes:
- There is a known issue with the use of [Generator Templates](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Template/#generator-templates) on the [List Generator](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-List/) on ArgoCD `2.7.0` -> `2.7.2` as a result of a breaking change in [this PR](https://github.com/argoproj/argo-cd/pull/12490). I am currently waiting on [the fix](https://github.com/argoproj/argo-cd/pull/13335) to be [merged into the `2.7` release](https://github.com/argoproj/argo-cd/pull/13508). I will leave this PR in draft until the fix has been released; however, I figured I might as well open it up for initial review.

oboukili commented on argocd/utils.go:
```diff
@@ -63,6 +63,16 @@ func expandStringList(l []interface{}) (result []string) {
 	return
 }
 
+func sliceOfString(slice []interface{}) []string {
+	result := make([]string, len(slice))
+
+	for i, s := range slice {
+		result[i] = s.(string)
+	}
+
+	return result
+}
+
```
Panics may happen, better be safe than sorry!

```suggestion
func sliceOfString(slice []interface{}) ([]string, error) {
	result := make([]string, len(slice))

	for i, s := range slice {
		switch s.(type) {
	        case string:
			result[i]= s.(string)
	        default:
	        	return nil, fmt.Errorf("element %v was not of type string: %T", s, s)
	        }
	}

	return result, nil
}

```

oboukili reviewed (changes requested):
Unbelievable work. To be honest, I can't even imagine the amount of time you put in...
I have to admit I blasted through the tests, as they are quite overwhelming, I agree with your decision to warn users beforehand, and encourage them to report bugs.

Had a few remarks, but the most important one that warranted a "request changes" was to discuss about the upsert within the resource creation.

Cheers,

onematchfox commented on argocd/resource_argocd_application_set.go:
```diff
@@ -0,0 +1,322 @@
+		}
+	}
+
+	featureApplicationSetSupported, err := si.isFeatureSupported(featureApplicationSet)
+	if err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  "feature not supported",
+				Detail:   err.Error(),
+			},
+		}
+	} else if !featureApplicationSetSupported {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary: fmt.Sprintf(
+					"application set is only supported from ArgoCD %s onwards",
+					featureVersionConstraintsMap[featureApplicationSet].String()),
+			},
+		}
+	}
```
Yeah, I think that would be a nice addition.

With regards to the feature "flags" in general, I have been playing around with the following ideas:
- Removing all flags for unsupported versions to clean up the codebase
- Removing the `error` returned by `isFeatureSupported`. The error is ultimately a "developer" error. I.e. passing some random/unknown value to an internal function. So, instead of returning an `error`, we could simply return false (unsupported) for unknown values which would then, assumedly, result in tests failing. Then all of these checks can be simplified down to something like

``` go
if !si.isFeatureSupported(featureApplicationSetSupported) {
	return []diag.Diagnostic{
		{
			Severity: diag.Error,
			Summary: fmt.Sprintf(
				"application set is only supported from ArgoCD %s onwards",
				featureVersionConstraintsMap[featureApplicationSet].String()),
		},
	}
}
```

or even better... combining this with your idea of "an error-to-diag converter" to provide an even terser syntax.

oboukili replied:
Absolutely 👍

onematchfox commented on argocd/structure_application_set.go:
```diff
@@ -0,0 +1,1472 @@
+package argocd
+
+import (
+	"encoding/json"
```
Nice. I wasn't aware of that project. Looks great. Will leave this as is for now since we use `encoding/json` in other places, so we can swap out the modules across the board in a separate PR.

onematchfox commented:
@oboukili Thanks for taking the time to review and for the positive comment. Yes, it did take quite a while (far longer than I initially bargained on), but luckily I have had quite a bit of time to kill recently.

I think I have addressed all comments (happy to discuss further if you disagree with me on anything).
